lib/ansible/plugins/connections/ssh.py

```diff
@@ -339,192 +286,395 @@ def _exec_command(self, cmd, tmp_path, in_data=None, sudoable=True):
+        (returncode, stdout, stderr) = self._run(cmd, in_data)
 
-            if passprompt:
-                self._display.debug("Sending privilege escalation password.")
-                stdin.write(self._play_context.become_pass + '\n')
-            else:
-                no_prompt_out = become_output
-                no_prompt_err = become_errput
+        if returncode != 0:
+            raise AnsibleError("failed to transfer file from {0}:\n{1}\n{2}".format(in_path, stdout, stderr))
 
+    def _run(self, cmd, in_data, sudoable=True):
+        '''
+        Starts the command and communicates with it until it ends.
+        '''
 
-        (returncode, stdout, stderr) = self._communicate(p, stdin, in_data, sudoable=sudoable)
+        self._display.vvv('SSH: EXEC {0}'.format(' '.join(map(pipes.quote,cmd))), host=self.host)
 
-        self.lock_host_keys(False)
+        # We begin by acquiring the connection lock, so that if ssh issues a
+        # host key verification prompt, output from other connections doesn't
+        # obscure it. The lock can be released once we know that the host key
+        # has been accepted (either because of known_hosts or the prompt).
 
-        controlpersisterror = 'Bad configuration option: ControlPersist' in stderr or 'unknown configuration option: ControlPersist' in stderr
+        self.connection_lock()
+        self._unlocked = False
 
-        if C.HOST_KEY_CHECKING:
-            if ssh_cmd[0] == "sshpass" and p.returncode == 6:
-                raise AnsibleError('Using a SSH password instead of a key is not possible because Host Key checking is enabled and sshpass does not support this.  Please add this host\'s fingerprint to your known_hosts file to manage this host.')
+        # Start the given command. If we don't need to pipeline data, we can try
+        # to use a pseudo-tty (ssh will have been invoked with -tt). If we are
+        # pipelining data, or can't create a pty, we fall back to using plain
+        # old pipes.
 
-        if p.returncode != 0 and controlpersisterror:
-            raise AnsibleError('using -c ssh on certain older ssh versions may not support ControlPersist, set ANSIBLE_SSH_ARGS="" (or ssh_args in [ssh_connection] section of the config file) before running again')
-        # FIXME: module name isn't in runner
-        #if p.returncode == 255 and (in_data or self.runner.module_name == 'raw'):
-        if p.returncode == 255 and in_data:
-            raise AnsibleConnectionFailure('SSH Error: data could not be sent to the remote host. Make sure this host can be reached over ssh')
+        p = None
+        if not in_data:
+            try:
+                # Make sure stdin is a proper pty to avoid tcgetattr errors
+                master, slave = pty.openpty()
+                p = subprocess.Popen(cmd, stdin=slave, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+                stdin = os.fdopen(master, 'w', 0)
+                os.close(slave)
+            except:
```

Comment:
A bare `except:` clause is usually frowned upon in Python land, because it tends to catch things like `KeyboardInterrupt` that you don't want to catch (and also hides any bugs that shouldn't be hidden).

Wouldn't `except (OSError, IOError):` suffice here?